### PR TITLE
Partial Matching, Refactor

### LINKED ISSUE #9: Suggest Fabrikate Components for Partial Match Deployments
As a user I want deployments that partially match Fabrikate components to be reflected in the component.yaml.

### Partial Match
A partial match is a scenario where a deployment on the cluster exists in a component, but the component contains one or more deployments that aren't present on my cluster.
#### Example
Cluster:         Prometheus
Component: Prometheus-Grafana

### Potential Solutions

1.  Allow the user to automatically use the more extensive component. The added functionality of the other deployments may prove to be useful for their cluster. A flag can be set that includes the partial match components in the component.yaml in a ready to use state with Fabrikate.

2. List the partial match components under a Partial Match comment header.  Include with each component the deployments that partially matched the cluster, giving the user a better idea of what's on the cluster.

### Chosen Solution

A combination of potential solutions 1 and 2. 

- [x] Identify Partial Match Deployments on the Cluster

- [x] Create List of Fabrikate Components that Partially Match

- [x] Add each item to a labelled "Partial Match" section with comments on next steps

- [x] Update Tests to reflect changes

### PR DESCRIPTION
**Pull request by andrewDoing**
closes #9 

Hydrate now determines partial match components, listing them in the component.yaml. In this scenario the cluster has prometheus, so Hydrate suggests prometheus-grafana.
Also note the inline comment stating the components found on the cluster that led to the partial match.
```yaml
subcomponents:
  # Partial Match Components
  - name: prometheus-grafana  # Cluster Components: prometheus
    source: https://github.com/microsoft/fabrikate-definitions
    method: git
    path: definitions/fabrikate-prometheus-grafana
```

A lot of refactoring is included as well, ideally it makes Hydrate easier to extend in the future.

### CODE REVIEW

File: hydrate/match.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            repo_components: list of Components from the definitions repository
+            full_match_leftovers: list of leftover deployments after get_full_matches
+
+        Returns:
+            PartialMatchResults: namedtuple
+                partial_matches: dict(key: component.name, value: list of components)
+                leftovers: list of components
+
+        """
+        PartialMatchResults = namedtuple('PartialMatchResults', ['partial_matches', 'leftovers'])
```
Comment (edaena) [inline hydrate/match.py]: leftovers -> no_matches?

File: hydrate/match.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+                    Match(category=MatchCategory.FULL_MATCH,
+                          cluster_matches=cluster_matches,
+                          repo_match=rc))
+
+        if cluster_set:
+            leftovers = [cc for cc in cluster_components if cc.name in cluster_set]
+
+        return FullMatchResults(full_matches, leftovers)
+
+    def get_partial_matches(self, full_match_leftovers):
```
Comment (edaena) [inline hydrate/match.py:108]: I feel like you can merge get_partial_matches and get_full_matches and just take the match category as argument: category=MatchCategory.FULL_MATCH
get_matches(category=MatchCategory.FULL_MATCH) have one by default if they pass partial then do the partial
those two methods look very similar

File: hydrate/match.py (hunk 1 above)
Comment (andrewDoing) [reply hydrate/match.py]: Yes, the leftovers get passed into the get_no_matches method. I suppose I could rename the leftovers to no_matches.

File: hydrate/match.py (hunk 2 above)
Comment (andrewDoing) [reply hydrate/match.py:108]: The partial matching method relies on the results of the full match method, so they need to be called in order. This is a good idea, I see what you mean that they are very similar... there's definitely repeated code.

**Review by andrebriggs — COMMENTED** [review]
Changes in `README.md` and `azure-pipelines.yml` seem to be noise as they are from a a bad merge. In the future please clean this up

File: tests/test_match.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    def test_init(self):
+        """Test the __init__ method."""
+        assert self.tst_matcher.repo_components == self.tst_repo_components
+
+    def test_match_components(self):
+        """Test the match_components method."""
+        pass
+
+    param_get_start_index = [[1, 1, 2], [-1, -1, -2], [0, 0, 0]]
```
Comment (andrebriggs) [inline tests/test_match.py:47]: @andrewDoing what does this mean?

**Review by andrebriggs — COMMENTED** [review]
This PR mixes major refactoring with adding new functionality. From the point of view of a reviewer it is hard to follow. Consider making smaller check ins. For instance: Add the match.py class file and they corresponding unit test file first as a PR.

File: hydrate/match.py (hunk 1 above)
Comment (andrewDoing) [reply hydrate/match.py]: Resolved in commit https://github.com/microsoft/hydrate/pull/26/commits/02df3ef98280187f991c39280c01e81c2ee9cc2a

**Review by andrebriggs — APPROVED** [review]
@andrewDoing approving this but only because the 16 files changed is too much for reviewers to really pay attention to. Please make smaller PRs or multiple PR for issues cc @edaena @madelineliao @samiyaakhtar
